### ISSUE TITLE
Sharding question with varying video lengths/resolutions

### ISSUE THREAD

**Issue by SiftingSands** [NVIDIA/DALI#3151]
If I have two videos, one with 1000 frames and the other with 10,000 frames, does DALI give each device the same number of frames/batches to ingest? I've noticed that all devices get evenly loaded at the beginning of the pipeline's run, but then only one device seems to be running after some time and stays that way until the end. My hunch was that if I had 2 GPUs and 2 videos, then each GPU was only handling frames from a single video. Is that the case?

I tried looking through the docs and examples (mainly https://docs.nvidia.com/deeplearning/dali/main-user-guide/docs/advanced_topics_sharding.html), but couldn't find an obvious answer. The pipeline that I'm using is wrapped around `pytorch.DALIGenericIterator`.

Thanks!

**Comment by klecki** [comment]
Hi @SiftingSands,
the Video Reader first prepares a collection of all possible sequences to be read (file, starting frame, length, etc), which is later optionally shuffled. This will describe all samples you can get from your two videos, there will be more samples obtained from the longer video. 

Next it traverses a selected shard (contiguous subset of those sequences), which you can se here: https://github.com/NVIDIA/DALI/blob/main/dali/operators/reader/loader/video_loader.cc#L734
(Notice the MoveToNextShard call). So the sharding between devices should be balanced, maybe with a difference of 1 (as there can be odd number of possible samples and you have 2 GPUs).

Do the videos differ in resolution for example? 

I recommend to check the profile and see if this is indeed Video Reader that introduces the difference. As a simple check you can also try running with synthetic data (so return a dummy tensor of desired size instead of running DALI preprocessing and see if it affects the speed of the training).

**Comment by SiftingSands** [comment]
Thanks for confirming that each device should be balanced by all possible sequences along the entire dataset. I did have different video resolutions, but there was a resizing op in the pipeline.

I'll have to check things on a minimal example, and see if I can reproduce my observations.

**Comment by SiftingSands** [comment]
After some digging, I did find some settings in my pipeline that reduced video loading performance (my fault). However, I still tried to produce a minimal example on synthetically created videos, and I unfortunately see unbalanced GPU utilization. Admittedly, my profiling consists of `tqdm` to count the iterations and `nvidia-smi` to track hardware utilization by eye.

Here's my minimal example : https://gist.github.com/SiftingSands/e7abad6ff18193ff10a5296412fab436
I would upload the videos that I used, but they're very large (>1GB). They consist of white noise, so data compression is infeasible. 
Here's the generation script : https://gist.github.com/SiftingSands/6eb1a17ee3f48a49d19133d82acce632

Environment info:
nvidia-dali-weekly-cuda110 : 1.4.0.dev20210704
2x Quadro P5000 (Haven't tested this out for similar behavior on other GPUs such as V/A100s)
NVIDIA-SMI : Driver Version: 465.19.01    CUDA Version: 11.3
torch : 1.9.0+cu111 (probably irrelevant)

I observe that gpu 1 (not sure if that'll be the same on your systems) finishes while gpu 0 is only ~halfway complete with its 57 batches. GPU 1 then sits idle until GPU 0 traverses all batches. Hopefully, those gists work fine on your end, and let me know if I misused some DALI settings responsible for this behavior.

**Comment by klecki** [comment]
Yes, Video Reader generates all the possible sequences, in your case consisting of 32 frames, for every video, grouping them one after another. 
Especially in the case of this repro, when the `n_frames` is the same, and there is no shuffling, you basically ended up with splitting your dataset in half (so the sequences from smaller video ended up on one GPU and from the bigger on the other).

As there is difference in resolution, this is probably where the difference comes from - the speed of NVDEC that we use underneath is resolution dependent.
If you run those pipelines in two threads without syncing them, the one with smaller data to process would indeed be expected to finish faster.

If you are worried about balancing the data, I could maybe suggest to either cut the video into smaller fragments and interleave them, or you could do this by preparing appropriate `file_list` (look at the argument of [readers.video](https://docs.nvidia.com/deeplearning/dali/main-user-guide/docs/supported_ops.html#nvidia.dali.fn.readers.video) for details), so the shards have a mix of both resolutions, for example 
```
small_res_file <label> 0 5000
big_res_file <label> 0 5000
small_res_file <label> 5000 10000
big_res_file <label> 5000 10000
```
(assuming both videos have 10k frames here).

But I would do that only if the difference in the pipelines speed is actual observable bottleneck. I think that switching up the resolution a lot might also have some overhead due to decoder reconfiguration (in my example it happens once so it shouldn't be a problem).

**Comment by SiftingSands** [comment]
Thanks for the detailed response! You mentioned syncing the two threads, is there built-in functionality to do that? I looked at https://docs.nvidia.com/deeplearning/dali/main-user-guide/docs/pipeline.html?highlight=sync#nvidia.dali.Pipeline but only saw `exec_async` and that didn't look like a solution.

**Comment by klecki** [comment]
I just meant that if you run two data augmentation pipelines in two independent threads (without the NN training), the pipeline that processes the smaller data will be faster and finish earlier as they are not constrained by anything other than the amount of data to process.

Usually you run a training loop, (DALI by default prepares next batches ahead of time, this is the `exec_async` and it's best to leave it set to True for optimal performance), and only when current iteration is being consumed by the network, the next iteration of DALI pipeline is scheduled for execution. So this is kind of natural "syncing point", and this is all I meant.

If you use the dedicated framework iterator from DALI, all the scheduling and controlling that is handled for you.

**Comment by SiftingSands** [comment]
I think that answers everything for now. Along the lines of mixing resolutions on the shards, setting `random_shuffle=True` in `fn.readers.video` seems to naturally load balance. I didn't see duplicate frames either.
